core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1844,6 +1900,68 @@ private Optional<OutputDataSizeEstimate> getEstimatedOutputDataSize()
             return Optional.of(new OutputDataSizeEstimate(estimateBuilder.build()));
         }
 
+        private Optional<OutputDataSizeEstimate> getEstimatedSmallStageOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup)
+        {
+            if (!smallStageEstimationEnabled) {
+                return Optional.empty();
+            }
+
+            if (smallStageRequireNoMorePartitions && !isNoMorePartitions()) {
+                return Optional.empty();
+            }
+
+            long currentOutputDataSize = 0;
+            for (long partitionOutputDataSize : outputDataSize) {
+                currentOutputDataSize += partitionOutputDataSize;
+            }
+            if (currentOutputDataSize > smallStageEstimationThreshold.toBytes()) {
+                // our output too big already
+                return Optional.empty();
+            }
+
+            List<StageInfo> sourceStages = this.getStageInfo().getSubStages();
+            long inputSizeEstimate = 0;
+            if (sourceStages.size() > 0) {
+                for (StageInfo sourceStageInfo : sourceStages) {
+                    StageId sourceStageId = sourceStageInfo.getStageId();
+                    StageExecution sourceStage = stageExecutionLookup.apply(sourceStageId);
+                    requireNonNull(sourceStage, "sourceStage is null");
+                    Optional<OutputDataSizeEstimate> sourceStageOutputDataSize = sourceStage.getOutputDataSize(stageExecutionLookup);
+
+                    if (sourceStageOutputDataSize.isEmpty()) {
+                        // cant estimate size of one of sources; should not happen in practice
+                        return Optional.empty();
+                    }
+
+                    inputSizeEstimate += sourceStageOutputDataSize.orElseThrow().getTotalSizeInBytes();
+                }
+                // TODO take table scans into account.
```

Comment:
Can you elaborate here? Table scans are not included in the calculation here?

Reply:
good point - it is not good I think.
if we have stage which reads from some remote stages and from a table, then TS will be ignored and we will estimate just on remote stages. This may be very inaccurate as TS can actually be huge. I will change this.